presto-main/src/main/java/com/facebook/presto/tracing/SimpleTracer.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static com.facebook.presto.spi.StandardErrorCode.DISTRIBUTED_TRACING_ERROR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class SimpleTracer
+        implements Tracer
+{
+    public final Map<String, SimpleTracerBlock> blockMap = new ConcurrentHashMap<>();
+    public final Map<String, SimpleTracerBlock> recorderBlockMap = new LinkedHashMap<>();
+    public final List<SimpleTracerPoint> pointList = new CopyOnWriteArrayList<>();
+
+    public SimpleTracer()
+    {
+        addPoint("Start tracing");
+    }
+
+    @Override
+    public void addPoint(String annotation)
+    {
+        pointList.add(new SimpleTracerPoint(annotation));
+    }
+
+    @Override
+    public void startBlock(String blockName, String annotation)
+    {
+        if (blockMap.containsKey(blockName)) {
+            throw new PrestoException(DISTRIBUTED_TRACING_ERROR, "Duplicated block inserted: " + blockName);
+        }
+        SimpleTracerBlock block = new SimpleTracerBlock(annotation);
+        blockMap.put(blockName, block);
+        synchronized (recorderBlockMap) {
+            recorderBlockMap.put(blockName, block);
+        }
+    }
+
+    @Override
+    public void addPointToBlock(String blockName, String annotation)
+    {
+        if (!blockMap.containsKey(blockName)) {
+            throw new PrestoException(DISTRIBUTED_TRACING_ERROR, "Adding point to non-existing block: " + blockName);
+        }
+        blockMap.get(blockName).addPoint(new SimpleTracerPoint(annotation));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **62 and 65**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **74**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).